src/model.ts

```diff
@@ -1,3 +1,44 @@
-export interface User {
-  name: string;
+export interface ValidationResult {
+  key?: string;
+  type: string;
+  succeeded: boolean;
+  errorMessage: string;
+}
+
+export const createDefaultValidationResult = (): ValidationResult => ({
+  key: '',
+  type: '',
+  succeeded: true,
+  errorMessage: '',
+});
+
+// Pending on ValidationEngine
+type ValidationResultSyncAsync = ValidationResult | Promise<ValidationResult>;
+
+export type FieldValidationFunction = (
+  value: any,
+  values: any,
+  customParams: any
+) => Promise<ValidationResult>;
+
+// renamoe to FullFieldValidation
+export interface FullFieldValidation {
+  validator: FieldValidationFunction;
```

Comment:
If the name of this parameter is `validator` maybe we could name the interface as well. What do you think?

Reply:
Not sure about this, the interface is providing the context to the member variable, I know it's a fieldValidation because it's enclosed inside FullFieldValidation.

Maybe we could rename this to something like:

```typescript
export interface FullFieldValidation {
  validationFn : FieldValidationFunction;
 // ... ?
```